packages/cdktf-cli/lib/cdktf-project.ts

```diff
@@ -82,18 +83,45 @@ function getMultipleStacks(
 
 // Returns the first stack that has no unmet dependencies
 // An unmet dependency is a dependency that has not been deployed yet
-function getStackWithNoUnmetDependencies(
+// If there is no unfinished stack, returns undefined
+// If there is no stack ready to be worked on, it returns a promise
+export async function getStackWithNoUnmetDependencies(
   stackExecutors: CdktfStack[]
-): CdktfStack | undefined {
-  return stackExecutors
-    .filter((executor) => executor.isPending)
-    .find((executor) =>
-      executor.stack.dependencies.every(
-        (dependency) =>
-          stackExecutors.find((executor) => executor.stack.name === dependency)
-            ?.currentState === "done"
-      )
-    );
+): Promise<CdktfStack | undefined> {
+  const pendingStacks = stackExecutors.filter((executor) => executor.isPending);
+
+  if (pendingStacks.length === 0) {
+    return undefined;
+  }
+
+  const currentlyReadyStack = pendingStacks.find((executor) =>
+    executor.stack.dependencies.every(
+      (dependency) =>
+        stackExecutors.find((executor) => executor.stack.name === dependency)
+          ?.currentState === "done"
+    )
+  );
+
+  if (currentlyReadyStack) {
+    return currentlyReadyStack;
+  }
+
+  const stackExecutionPromises = stackExecutors.reduce((carry, executor) => {
+    const promise = executor.currentWorkPromise;
+    if (promise) {
+      return [...carry, promise];
+    } else {
+      return carry;
+    }
+  }, [] as Promise<void>[]);
+
+  if (!stackExecutionPromises.length) {
+    return undefined;
+  }
+
+  await Promise.race(stackExecutionPromises);
+
+  return await getStackWithNoUnmetDependencies(stackExecutors);
```

Comment:
> Another short question: What happens if a stack is waiting for user approval? Will its currentWorkPromise be resolved at that point (which might cause a loop of this function in that case; because it is partially done but not currentState === "done")?

The currentWorkPromise is resolved when the state machine is in the done state, it will not resolve if you e.g. approve. if you dismiss it will resolve, but at that point all the executors what depended on this stack will have been removed from the stacksToRun list

Reply:
Ah, alright 👍 Sounds safe enough to me.
I'd still add a tracelog message because that might really save us if we break the statemachine or something like that and it wouldn't hurt if not :)